### PR TITLE
reduce compile time by making output label generation non-generic

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

Split `HasOutputTypeLabel::output_type_label` method implementation non-generic part into a separate function. This significantly reduces compile time without risking any performance regressions. That function is currently only used for debug network visualization using graphviz, but still contributed a significant compilation time.

I've made a single attempt to profile the compiler itself, and it turned out that the compiler spent a significant amount of time trying to resolve `Pattern` implementation for closures in that method. Each of those also had to separately go through codegen and optimization. That happened for each node type in the codebase, per crate. Moving that into separate non-inline function removed all those unnecessary duplicates from. I also took this opportunity to rewrite that small piece of parsing to make it a bit cleaner.

The method of measurement is explained here: https://blog.rust-lang.org/inside-rust/2020/02/25/intro-rustc-self-profile.html
In this specific case, I've used `self-profile` to generate a profile for all crates, then used `crox` and [perfetto](https://ui.perfetto.dev/) to analyze the output.

This also shows a pattern to be aware of - using closures in generic context forces the compiler to make a separate closure type per each instantiation, even if that closure doesn't close on anything and could be a static function. This in effect forces even more instantiations or unnecessary type resolutions for all code paths that touch that closure. Using static functions or separating the non-generic part away in those cases would likely continue to help with compile times and file size.

## Comparison

The measurement was done on same machine under same environment, cleaning the build artifacts inbetween runs.

| | before |  after |
|-|-|-|
|total build time|5m 5.5s|3m 59.0s|
| `ide-view-graph-editor` crate build time| 85.68s | 49.73s |
| `ide-view-component-list-panel-grid` crate build time | 49.88s | 32.07s |
| `ide-view` crate build time | 29.05s | 17.5s |
| `enso_gui.wasm` file size before wasm-opt | 83.6 MB | 80.9 MB |
| `enso_gui.wasm` file size after wasm-opt | 67.3 MB | 65.3 MB |

image

	
### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

**Review by wdanilo — APPROVED** [review]

**Comment by wdanilo** [comment]
Amazing job with this!

File: lib/rust/frp/src/stream.rs [hunk 1]
```diff
@@ -659,24 +659,24 @@ where Def: InputBehaviors
...
-            label = &label[0..label.len() - sfx.len()];
-        }
-        label.into()
+    fn output_type_label(&self) -> Label {
+        type_name_to_output_label(type_name::<Def>())
     }
 }
 
+// The label transformation logic is a separate non-generic function, so that it can be compiled
+// only once for all node types. This has a noticeable impact on compilation time.
```
Comment (wdanilo) [inline lib/rust/frp/src/stream.rs:671]: add here a link to this PR if it mentions "noticeable impact on compile time" - it's worth proving statements :)

**Comment by wdanilo** [comment]
Also, an idea - if it affects time so much, maybe its worth making a small macro for extracting functions this way? We have many places where it might help a lot.

**Comment by wdanilo** [comment]
One more thing - if this makes the total binary size smaller, please change it in our build scripts (we are checking if the binary file is not too big there)

**Review by mwu-tow — APPROVED** [review]
👏

**Pull request merged**